src/Functions/printf.cpp

```diff
@@ -0,0 +1,309 @@
+
+        String format = c0_const_string->getValue<String>();
+        auto instructions = buildInstructions(format, arguments, input_rows_count);
+
+        ColumnsWithTypeAndName concat_args(instructions.size());
+        for (size_t i = 0; i < instructions.size(); ++i)
+        {
+            // std::cout << "instruction[" << i << "]:" << instructions[i].toString() << std::endl;
+            concat_args[i] = instructions[i].execute();
+            // std::cout << "concat_args[" << i << "]:" << concat_args[i].dumpStructure() << std::endl;
+        }
+
+        auto res = function_concat->build(concat_args)->execute(concat_args, std::make_shared<DataTypeString>(), input_rows_count);
+        return res;
+    }
+
+private:
+    std::vector<Instruction>
+    buildInstructions(const String & format, const ColumnsWithTypeAndName & arguments, size_t input_rows_count) const
+    {
+        std::vector<Instruction> instructions;
+        instructions.reserve(arguments.size());
+
+        auto append_instruction = [&](const char * begin, const char * end, const ColumnWithTypeAndName & arg)
+        {
+            Instruction instr;
+            instr.rows = input_rows_count;
+            instr.format = std::string_view(begin, end - begin);
+
+            size_t size = end - begin;
+            if (size > 1 && begin[0] == '%' and begin[1] != '%')
+            {
+                instr.is_literal = false;
+                instr.input = arg;
+            }
+            else
+            {
+                instr.is_literal = true;
+            }
+            instructions.emplace_back(std::move(instr));
+        };
+
+        auto check_index_range = [&](size_t idx)
+        {
+            if (idx >= arguments.size())
+                throw Exception(
+                    ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                    "Number of arguments for function {} doesn't match: passed {}, but format is {}",
+                    getName(),
+                    arguments.size(),
+                    format);
+        };
+
+        const char * begin = format.data();
+        const char * end = format.data() + format.size();
+        const char * curr = begin;
+        size_t idx = 0;
+        while (curr < end)
+        {
+            const char * tmp = curr;
+            bool is_first = curr == begin; /// If current instruction is the first one
+            bool is_literal = false; /// If current instruction is literal string without any argument
+            if (is_first)
+            {
+                if (*curr != '%')
+                    is_literal = true;
+                else if (curr + 1 < end && *(curr + 1) == '%')
+                    is_literal = true;
+                else
+                    ++idx; /// Skip first argument if first instruction is not literal
+            }
+
+            if (!is_literal)
+                ++curr;
+
+            while (curr < end)
```

Comment:
What the motivation behind splitting each format arg and execute it separately instead of calling `fmt::sptrintf` for the whole string?
Though it totally makes sense to pre-validate input string not to rely on `fmt::sptrintf` completely. What safety guaranties `fmt::sptrintf` makes, by the way? Is it safe to call for arbitrary user input?

Reply:
Because fmt::sprintf only receives variable length arguments. It is impossible to get exact arugment number during compiling.